File: week 1/connected_components/connected_components.py
def number_of_components(adj):
    result = 0
    #write your code here
    visited = [0] * len(adj)
    for i in xrange(len(adj)):
        if not visited[i]:
            explore(adj, i, visited)
            result += 1
    return result

def explore(adj, x, visited):
    visited[x] = 1
    for i in xrange(len(adj[x])):
        if not visited[adj[x][i]]:
            explore(adj, adj[x][i], visited)

if __name__ == '__main__':
    n, m = map(int, raw_input().split())
    adj = [[] for _ in range(n)]
    for i in range(m):
        a, b = map(int, raw_input().split())
        # adjacency list		
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    print(number_of_components(adj))

File: week 1/connected_components/ConnectedComponents.java
import java.util.ArrayList;
import java.util.Scanner;

public class ConnectedComponents {
    private static int numberOfComponents(ArrayList<Integer>[] adj) {
        int result = 0;
		//write your code here
		int[] visited = new int[adj.length];
		for (int i = 0; i < adj.length; i++) {
			if (visited[i] == 0) {
				explore(adj, i, visited);
                result++;
			}
		}
        return result;
    }
	
	private static void explore(ArrayList<Integer>[] adj, int x, int[] visited) {
		visited[x] = 1;
		for (int i = 0; i < adj[x].size(); i++) {
			if (visited[adj[x].get(i)] == 0)
				explore(adj, adj[x].get(i), visited);
		}
	}

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int n = scanner.nextInt();
        int m = scanner.nextInt();
        ArrayList<Integer>[] adj = (ArrayList<Integer>[])new ArrayList[n];
        for (int i = 0; i < n; i++) {
            adj[i] = new ArrayList<Integer>();
        }
        for (int i = 0; i < m; i++) {
            int x, y;
            x = scanner.nextInt();
            y = scanner.nextInt();
            adj[x - 1].add(y - 1);
            adj[y - 1].add(x - 1);
        }
        System.out.println(numberOfComponents(adj));
    }
}

